neuralmonkey/decoders/sequence_labeler.py

```diff
@@ -52,64 +62,36 @@ def target_tokens(self) -> tf.Tensor:
 
     @tensor
     def train_targets(self) -> tf.Tensor:
-        return self.vocabulary.strings_to_indices(self.target_tokens)
+        return self.vocabulary.strings_to_indices(
+            self.dataset[self.data_id])
 
     @tensor
     def train_mask(self) -> tf.Tensor:
         return sentence_mask(self.train_targets)
 
-    @property
-    def rnn_size(self) -> int:
-        return int(self.encoder.temporal_states.get_shape()[-1])
-
-    @tensor
-    def decoding_w(self) -> tf.Variable:
-        return get_variable(
-            name="state_to_word_W",
-            shape=[self.rnn_size, len(self.vocabulary)])
-
     @tensor
-    def decoding_b(self) -> tf.Variable:
-        return get_variable(
-            name="state_to_word_b",
-            shape=[len(self.vocabulary)],
-            initializer=tf.zeros_initializer())
+    def concatenated_inputs(self) -> tf.Tensor:
+        return tf.concat(
+            [inp.temporal_states for inp in self.encoders], axis=2)
 
     @tensor
-    def decoding_residual_w(self) -> tf.Variable:
-        input_dim = self.encoder.input_sequence.dimension
-        return get_variable(
-            name="emb_to_word_W",
-            shape=[input_dim, len(self.vocabulary)])
+    def states(self) -> tf.Tensor:
+        states = dropout(
+            self.concatenated_inputs, self.dropout_keep_prob, self.train_mode)
+
+        if self.hidden_dim is not None:
+            hidden = tf.layers.dense(
+                states, self.hidden_dim, self.activation,
+                name="hidden_layer")
+            # pylint: disable=redefined-variable-type
+            states = dropout(hidden, self.dropout_keep_prob, self.train_mode)
+            # pylint: enable=redefined-variable-type
+        return states
 
     @tensor
     def logits(self) -> tf.Tensor:
-        # To multiply 3-D matrix (encoder hidden states) by a 2-D matrix
-        # (weights), we use 1-by-1 convolution (similar trick can be found in
-        # attention computation)
-
-        # TODO dropout needs to be revisited
-
-        encoder_states = tf.expand_dims(self.encoder.temporal_states, 2)
-        weights_4d = tf.expand_dims(tf.expand_dims(self.decoding_w, 0), 0)
-
-        multiplication = tf.nn.conv2d(
-            encoder_states, weights_4d, [1, 1, 1, 1], "SAME")
-        multiplication_3d = tf.squeeze(multiplication, axis=[2])
-
-        biases_3d = tf.expand_dims(tf.expand_dims(self.decoding_b, 0), 0)
-
-        embedded_inputs = tf.expand_dims(
-            self.encoder.input_sequence.temporal_states, 2)
-        dweights_4d = tf.expand_dims(
-            tf.expand_dims(self.decoding_residual_w, 0), 0)
-
-        dmultiplication = tf.nn.conv2d(
-            embedded_inputs, dweights_4d, [1, 1, 1, 1], "SAME")
-        dmultiplication_3d = tf.squeeze(dmultiplication, axis=[2])
-
-        logits = multiplication_3d + dmultiplication_3d + biases_3d
-        return logits
+        return tf.layers.dense(
```

Comment:
Dropout na self.states?

Reply:
jo vidim že tady je to teda opačně než by mělo - states by se měly asi vracet nedropovaný a tohle by si je mělo dropovat.